### PR TITLE
[core] Fix SIGABRT on erase call

### LINKED ISSUE #17080: [placeholder] Investigate Dask on Ray segfault from dlmalloc in nightly release tests
https://buildkite.com/ray-project/periodic-ci/builds/555#_

### PR DESCRIPTION
**Pull request by scv119**
we've seen SIGABRT EraseFromObjectTable call, which is most likely caused by `auto &object = store_info_.objects[object_id];` which creates an empty object entry if the object_id doesn't exist. This fixes the behavior.

also per dlmalloc's [documents](http://gee.cs.oswego.edu/pub/misc/malloc-2.8.4.c):

```
/*
  free(void* p)
  Releases the chunk of memory pointed to by p, that had been previously
  allocated using malloc or a related routine such as realloc.
  It has no effect if p is null. If p was not malloced or already
  freed, free(p) will by default cause the current program to abort.
*/
```

```
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299: *** SIGABRT received at time=1626467718 on cpu 7 ***
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299: PC: @     0x7fd426e7018b  (unknown)  raise
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x7fd4273873c0       3936  (unknown)
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4ab9e2         32  plasma::PlasmaAllocator::Free()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4b5fb4         96  plasma::PlasmaStore::EraseFromObjectTable()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4b6111         48  plasma::PlasmaStore::DeleteObject()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4bb925        448  plasma::PlasmaStore::ProcessMessage()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4ad2ff         32  std::_Function_handler<>::_M_invoke()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4ce966        160  std::_Function_handler<>::_M_invoke()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f7650f2        128  ray::ClientConnection::ProcessMessage()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f761c78        128  boost::asio::detail::read_op<>::operator()()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f76202b        128  boost::asio::detail::executor_function<>::do_complete()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f357880         80  boost::asio::io_context::executor_type::dispatch<>()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f762973        192  boost::asio::executor::dispatch<>()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f762b68        224  boost::asio::detail::reactive_socket_recv_op<>::do_complete()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854fb7d351        112  boost::asio::detail::scheduler::do_run_one()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854fb7d481        176  boost::asio::detail::scheduler::run()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854fb7f680         64  boost::asio::io_context::run()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f4ac41b        304  plasma::PlasmaStoreRunner::Start()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854f43ef67        208  std::thread::_State_impl<>::_M_run()
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @     0x55854fbc0960  (unknown)  execute_native_thread_routine
[2021-07-16 13:35:18,484 E 96 123] logging.cc:299:     @ ... and at least 3 more frames
```

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]

File: src/ray/object_manager/plasma/store.cc [hunk 1]
```diff
@@ -623,9 +623,16 @@ int PlasmaStore::RemoveFromClientObjectIds(const ObjectID &object_id,
 }
 
 void PlasmaStore::EraseFromObjectTable(const ObjectID &object_id) {
-  auto &object = store_info_.objects[object_id];
+  auto object = GetObjectTableEntry(&store_info_, object_id);
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:626]: Ah nice catch. I assume the symptom of this was we would try to free(nullptr) or similar?

**Review by ericl — APPROVED** [review]
Ah the default constructor strikes again

File: src/ray/object_manager/plasma/store.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc:626]: It looks like free nullptr is fine (according to the dmalloc doc string, it is no op), but calling get on the default unique ptr seems to have undefined behavior

Comment (scv119) [reply src/ray/object_manager/plasma/store.cc:626]: actually calling operator*/-> on default constructed a unique_ptr yields undefined behavior. So calling `PlasmaStore::Free(object->data...)` could potentially lead to free any address.



https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*

> operator* and operator-> provide access to the object owned by *this.
The behavior is undefined if get() == nullptr.

**Pull request merged**